### PR TITLE
fix: support legacy equality overrides

### PR DESCRIPTION
**Pull request by creativecreatorormaybenot**
## Status
**READY**

## Breaking Changes
NO

## Description

Fixes the equality override not allowing `Object?` anymore with the null safety update. This had the effect of not permitting to mixin equatable on classes with an existing, e.g. `dynamic` equality override.

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Examples

### CODE REVIEW

**Review by felangel — COMMENTED** [review]
Thanks so much for the contribution! Can you please add tests for these changes? Thanks 🙏

**Comment by creativecreatorormaybenot** [comment]
@felangel How do I test compile time errors?

**Comment by felangel** [comment]
@creativecreatorormaybenot can you provide a snippet of the usage you'd like to allow with this PR? I can take a closer look shortly 👍

**Comment by creativecreatorormaybenot** [comment]
```dart
class A {
  @override
  operator ==(dynamic other) {
     // ...
  }
}

class B extends A with EquatableMixin {}
```

This does not compile with null safety because EquatableMixin does not provide a valid override for `==`.

Also, I did not test this, what about `equatable == null`?

**Comment by felangel** [comment]
> ```dart
> class A {
>   @override
>   operator ==(dynamic other) {
>      // ...
>   }
> }
> 
> class B extends A with EquatableMixin {}
> ```
> 
> This does not compile with null safety because EquatableMixin does not provide a valid override for `==`.
> 
> Also, I did not test this, what about `equatable == null`?

I don't think I understand the use-case. You should replace `dynamic` with `Object` in the `==` operator override.

Take the following example:

```dart
class A {
  const A(this.x);

  final int x;

  @override
  bool operator ==(Object o) {
    if (identical(this, o)) return true;

    return o is A && o.x == x;
  }

  @override
  int get hashCode => x.hashCode;
}

class B extends A with EquatableMixin {
  const B(int x) : super(x);

  @override
  List<Object?> get props => [x];
}

void main() {
  const a = A(0);
  const b = B(0);

  print(a == null); // false
  print(b == null); // false
  print(a == b); // true

  A? nullableA;
  B? nullableB;

  print(a == nullableA); // false
  print(b == nullableB); // false

  nullableA = const A(0);
  nullableB = const B(0);

  print(a == nullableA); // true
  print(b == nullableB); // true
}
```

I'm not sure why you'd ever want to compare a non-nullable instance to `null` though when using sound null safety as it will never be `true`.

Sorry if I'm missing something! Let me know what you think 👍

**Comment by creativecreatorormaybenot** [comment]
@felangel Working with legacy code.

**Comment by creativecreatorormaybenot** [comment]
@felangel In legacy code, overrides often used `dynamic`. If we can argue that this will not happen after libraries are migrated, we can not make this change.

On the other hand, I do not think it does any harm, does it? As you check for `is Equatable` first, which will make flow analysis promote the type anyway.

File: lib/src/equatable.dart [hunk 1]
```diff
@@ -43,7 +43,7 @@ abstract class Equatable {
   bool? get stringify => null;
 
   @override
-  bool operator ==(Object other) =>
+  bool operator ==(Object? other) =>
```
Comment (felangel) [inline lib/src/equatable.dart:46]: why is this change needed in the case of `Equatable`? Since developers are in full control of this class I would prefer not to make this change unless there's a concrete use-case for it. Thoughts?

**Comment by felangel** [comment]
@creativecreatorormaybenot thanks for the clarification. I understand the use-case for `EquatableMixin` but I don't think it's necessary to apply to the `Equatable`. wdyt?

**Comment by felangel** [comment]
Just opened #108 to add support for this to `EquatableMixin` and started a discussion on #84.

**Comment by creativecreatorormaybenot** [comment]
@felangel What the ...?

This slightly offends me in terms of open source to be frank.

**Comment by creativecreatorormaybenot** [comment]
@felangel What the ...?

This slightly offends me in terms of open source to be frank.

---

(I do appreciate the effort, though 🚀)

**Comment by creativecreatorormaybenot** [comment]
@felangel The proper way would be asking the original author to make any necessary adjustments (or potentially picking their change) but not creating it again yourself.

**Comment by felangel** [comment]
> @felangel What the ...?
> 
> This slightly offends me in terms of open source to be frank.
> 
> (I do appreciate the effort, though 🚀)

I wasn't planning to merge that branch, just wanted to have a clean branch with the diff I was referring to. Can you rebase your changes as I wasn't able to force push to your branch.

**Comment by creativecreatorormaybenot** [comment]
@felangel I see, make sense. I suppose I am used to having discussion on PRs as well.

Anyway, I looked at your PR and I think the implementation of the tests is fine - there is no need to duplicate the work here. I will add a note to your PR.
